### ISSUE TITLE
Backend strategy

### ISSUE THREAD

**Issue by mayou36** [zfit/zfit#21]
We currently fully rely on TF. In the foreseeable future, this may change, so the backend should desirably be exchangeable. I came up with three possible ways to go, each with their own advantages and disadvantages.

## 1 Backend wrapping on low level (like Keras does)
we could in principle wrap tensorflow, pytorch on a low level and use this functions for operations (like `dot` instead of `tf.matmul` or `torch.mm` etc.).  

#### Advantage: 
- _once_ implemented, it yields full flexibility!

#### Disadvantage:
- a lot of work! Example: [keras wrapped tf backend](https://github.com/keras-team/keras/blob/master/keras/backend/tensorflow_backend.py)  
- yet another new API to learn for users already used to tf or torch

## 2 Strategy pattern
Simple: a pdf contains several implementations of the pdf (with tf, torch). Depending on the backend, a different implementation is chosen each time (of course, automatically).

#### Advantage:
- Allows easily even for mixed backends (in theory), allows flexible backend change
- Somehow beautiful solution :)
#### Disadvantage:
- each object has to be taken care of individually: this works nice for, say, pdfs, but get's fast out of hands when implementing `Variable` and does not work for static functions (or not so easily).

## 3 Name mangling on import time
Knowing the backend, on import time the right module is made available in the namespace. E.g. there is a `sum_tf`, `sum_torch` function and, depending on the backend chosen, either `sum = sum_tf` or `sum = sum_torch` leads to `from foo import sum` always yielding the right function.

#### Advantage:
- super easy, though rather brute force and inelegant implementation
#### Disadvantage:
- static: mixing not possible (or with backend knowledge required), flexible change not possible (first set backend, then do imports)

# Conclusion
1 is unrealistic. 2 can get out of hands with complexity. So realistically speaking, I think we will go for 3. The backend is not an important part and "could" even be just fixed. All we need to make sure is therefore to have

- [ ] a backend flag that can be set without other imports
- [ ] a function (or some kind of strategy) to mangle the names ("worst" case: simple ifs)

What are your thoughts about that @apuignav ?

**Comment by mayou36** [comment]
Issue moved to [zfit/zfit-developement #18](https://github.com/zfit/zfit-developement/issues/18) via [**ZenHub**](https://www.zenhub.com/)
